sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1281,9 +1294,107 @@ CleanupEventListFromResetCmdList(std::vector<pi_event> &EventListToCleanup,
+static pi_result CleanupEventsInImmCmdLists(pi_queue Queue,
+                                            bool QueueLocked = false,
+                                            bool QueueSynced = false,
+                                            pi_event CompletedEvent = nullptr) {
+  // Handle only immediate command lists here.
+  if (!Queue || !Queue->Device->useImmediateCommandLists())
+    return PI_SUCCESS;
+
+  std::vector<pi_event> EventListToCleanup;
+  {
+    std::unique_lock<pi_shared_mutex> QueueLock(Queue->Mutex, std::defer_lock);
+    if (!QueueLocked)
+      QueueLock.lock();
+    // If queue is locked and fully synchronized then cleanup all events.
+    // If queue is not locked then by this time there may be new submitted
+    // commands so we can't do full cleanup.
+    if (QueueLocked &&
+        (QueueSynced || (Queue->isInOrderQueue() &&
+                         (CompletedEvent == Queue->LastCommandEvent ||
+                          !Queue->LastCommandEvent)))) {
+      Queue->LastCommandEvent = nullptr;
+      for (auto &&It = Queue->CommandListMap.begin();
+           It != Queue->CommandListMap.end(); ++It) {
+        std::move(std::begin(It->second.EventList),
+                  std::end(It->second.EventList),
+                  std::back_inserter(EventListToCleanup));
+        It->second.EventList.clear();
+      }
+    } else if (Queue->isInOrderQueue() && CompletedEvent) {
+      // If the queue is in-order and we have information about completed event
+      // then cleanup all events preceding to CompletedEvent including itself.
+
+      // Check that the comleted event has associated command list.
+      if (!(CompletedEvent->CommandList &&
+            CompletedEvent->CommandList.value() != Queue->CommandListMap.end()))
+        return PI_SUCCESS;
+
+      auto &CmdListEvents =
+          CompletedEvent->CommandList.value()->second.EventList;
+      auto CompletedEventIt =
+          std::find(CmdListEvents.begin(), CmdListEvents.end(), CompletedEvent);
+      if (CompletedEventIt != CmdListEvents.end()) {
+        // Remember the first event in this command list, we know that it
+        // is completed.
+        auto FirstEvent = CmdListEvents.front();
+        // We can cleanup all events prior to the completed event.
+        std::move(std::begin(CmdListEvents), CompletedEventIt + 1,
+                  std::back_inserter(EventListToCleanup));
+        CmdListEvents.erase(CmdListEvents.begin(), CompletedEventIt + 1);
+        // We can cleanup events from all command lists preceding the command
+        // list containing completed event.
+        for (auto &&It = Queue->CommandListMap.begin();
+             It != Queue->CommandListMap.end(); ++It) {
+          if (!It->second.EventList.empty()) {
+            auto LastEventInCmdList = It->second.EventList.back();
+            // If the command lists' last event is in the waitlist of the first
+            // event in the command list with completed event.
```

Comment:
Why do you look at "last" to "first" dependency here? cannot it be "middle" to "middle", e.g. when switching between copy and compute lists with in-order, or changing compute lists for submitting to multiple engines?

Reply:
Hm, indeed, good concern. Even though it is not an error, it doesn't make sense to look  at "last" to "first" dependency because it is unlikely to be such case for immediate command lists. Then it looks like I should cleanup all prior events from the current command list only, do you agree? It doesn't look like I can effectively clean events from other command lists as well.

Reply:
Perhaps you could find out the "preceding" (in in-order bookkeeping sense) event and look for that in all the command-lists (if it is not the previous command in this command-list) and restart backward walk from there?

Reply:
Problem is that if I have only one command list (like it is now) then I will keep walking backwards in the same command list and will never find a "preceding" event from a different command list. And to find a "preceding" event I need to check waitlist every time. This approach seems too expensive.

Reply:
How about not the "preceding" in-order event by looking at first/last event in the wait list only? Then it is either the previous event in this command-list or you search for it in other command-lists.

Reply:
This may work, I will check that separately.